### ISSUE TITLE
Classmap issue after update to 2.1.7

### ISSUE THREAD

**Issue by Webonaute** [composer/composer#10102]
I just upgraded to composer 2.1.7 from 2.1.6 and composer classmap cache stopped working or there is some breaking change in new version.

any subsequent command have error like : 
`Noticed exception 'Error' with message 'Class '\Contract\V1\Config\GPBMetadata\Config' not found' in /var/www/project/vendor/someVendor/Contract/V1/Config/Config.php`

only change is upgrading composer to 2.1.7.

here is the command we use in our automations scripts:
`su -p -c "composer dump-autoload --optimize --classmap-authoritative --apcu" www-data`

**Comment by mitelg** [comment]
we are experiencing a similar issue in https://github.com/shopware/shopware

> Error: Class 'Smarty_Internal_Compile_Block' not found

are we doing something wrong in our `composer.json` ?

**Comment by mitelg** [comment]
okay, we found out that the issue occurs if the file contains two classes:

https://github.com/shopware/shopware/blob/5.7/engine/Library/Smarty/sysplugins/smarty_internal_compile_block.php

once we split it up in two files, it works correctly again.

**Comment by Seldaek** [comment]
Thanks @mitelg for the repro case with file. That one was due to the file having invalid unicode sequences. If I open it and save it again it starts working for me, which I guess explains why splitting it in two helped you.

Running a diff pre/post save shows:

```
104,105c104,105
<                             $search = array("!({$_ldl}{$al}block[\s\S]*?{$name}[\s\S]*?{$_rdl})([\s\S]*?)({$_ldl}{$al}\\\$smarty\.block\.child\s*{$_rdl})([\s\S]*?)({$_ldl}{$al}/block\s*{$_rdl})!", "/§§§child§§§/");
<                             $replace = array('\2§§§child§§§\4', $replacement);
---
>                             $search = array("!({$_ldl}{$al}block[\s\S]*?{$name}[\s\S]*?{$_rdl})([\s\S]*?)({$_ldl}{$al}\\\$smarty\.block\.child\s*{$_rdl})([\s\S]*?)({$_ldl}{$al}/block\s*{$_rdl})!", "/���child���/");
>                             $replace = array('\2���child���\4', $replacement);
```

So it looks like the � which as per https://www.babelstone.co.uk/Unicode/whatisit.html are apprently `U+FFFD : REPLACEMENT CHARACTER` aren't very much liked by my editor nor by PCRE. Anyway that should be fixed.

@Webonaute could you check if the snapshot build also fixes your issue? If not please share the source so I can try and repro/debug. You can try with `composer self-update --snapshot`. Thanks!

**Comment by mitelg** [comment]
oh, what a strange problem :see_no_evil: :smile: 

thank you so much for the fast fix! I tested the snapshot version, and can confirm it worked again for us :blush: :+1: :tada:

**Comment by astepin** [comment]
I can also confirm that the problem is fixed with the latest snapshot. It is quite easy to encounter this problem when developing with Windows.

**Comment by Seldaek** [comment]
https://github.com/composer/composer/releases/tag/2.1.8 is out with the fix.

**Comment by Webonaute** [comment]
Thank. I confirm issue is fixed with 2.1.8 for me.
